allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+        for article in dataset:
+            for paragraph_json in article["paragraphs"]:
+                context = paragraph_json["context"]
+                for question_answer in paragraph_json["qas"]:
+                    answers = [answer_json["text"] for answer_json in question_answer["answers"]]
+
+                    # This version does not support no-answer questions.
+                    if len(answers) <= 0:
+                        continue
+
+                    # Just like huggingface, we only use the first answer for training.
+                    first_answer_offset = int(question_answer["answers"][0]["answer_start"])
+
+                    instances = self.make_instances(
+                        question_answer.get("id", None),
+                        question_answer["question"],
+                        answers,
+                        context,
+                        first_answer_offset,
+                    )
+                    instances_yielded = 0
+                    for instance in instances:
+                        yield instance
+                        instances_yielded += 1
+                    if instances_yielded > 1:
+                        questions_with_more_than_one_instance += 1
+                    question_count += 1
+                    if self.max_questions is not None and question_count >= self.max_questions:
+                        return
+
+        logger.info(
+            "%d (%.2f%%) questions have more than one instance",
+            questions_with_more_than_one_instance,
+            100 * questions_with_more_than_one_instance / question_count,
+        )
+
+    def make_instances(
+        self, qid: str, question: str, answers: List[str], context: str, first_answer_offset: int
+    ) -> Iterable[Instance]:
+        answer_end = first_answer_offset + len(answers[0])
+
+        # tokenize context by spaces first, and then with the wordpiece tokenizer
+        # For RoBERTa, this produces a bug where every token is marked as beginning-of-sentence. The huggingface
+        # implementation has the same bug, so we're leaving it in.
+        tokenized_context = []
+        token_start = 0
+        for i, c in enumerate(context):
+            if c.isspace():
+                for wordpiece in self._tokenizer.tokenize(context[token_start:i]):
+                    if wordpiece.idx is None:
+                        tokenized_context.append(wordpiece)
+                    else:
+                        tokenized_context.append(
+                            wordpiece._replace(idx=wordpiece.idx + token_start)
+                        )
+                token_start = i + 1
+        for wordpiece in self._tokenizer.tokenize(context[token_start:]):
+            if wordpiece.idx is None:
+                tokenized_context.append(wordpiece)
+            else:
+                tokenized_context.append(wordpiece._replace(idx=wordpiece.idx + token_start))
+
+        token_answer_span_start = 0
```

Comment:
Don't we have a utility for this? https://github.com/allenai/allennlp-reading-comprehension/blob/fa60af5736a22455d275e663d3dd1ecc838e400c/allennlp_rc/dataset_readers/util.py#L47  Can this be re-used?

Reply:
`char_span_to_token_span()` does not deal with tokens that don't have an offset. This happens when we can't find the right offset. The workarounds are not great.

If we want to use it, it might be the best option to add dealing with no-offset tokens into `char_span_to_token_span()`.

Reply:
I would vote for extending the utility to handle this and sharing the code.

Reply:
172e3eeccaa830452c709a5199adc5a1e6810e13